client/services/code_preprocessors/PythonCodePreprocessorService.js

```diff
@@ -54,43 +93,66 @@ tie.factory('PythonCodePreprocessorService', [
         ].join('\n');
       });
 
       return testCode;
-    }
+    };
 
-    return {
-      // Wraps a WRAPPER_CLASS_NAME class around the series of functions in a
-      // given code snippet.
-      _wrapCodeIntoClass: function(code) {
-        var codeLines = code.trim().split('\n');
+    var generatePerformanceTestCode = function(performanceTests) {
+      performanceTests.forEach(function(test, index) {
+        var qualifiedEvaluationFunctionName = (
+          WRAPPER_CLASS_NAME + '().' + test.getEvaluationFunction());
+        var qualifiedTransformationFunctionName = (
+          WRAPPER_CLASS_NAME + '().' + test.getTransformationFunction());
+        // TODO(eyurko): Make this work for other runtimes.
+        // TODO(eyurko): 30 seems very scientific. Maybe stop hardcoding?
+        testCode += [
+          '',
+          'def get_test_input(atom, input_size):',
+          '    return ' + qualifiedTransformationFunctionName + '(atom, input_size)',
+          '',
+          'def run_performance_test(test_input):',
+          '    time_array = []',
+          '    for input_size in [10, 100]:',
+          '        start = time.time()',
+          '        output = ' + qualifiedEvaluationFunctionName + '(get_test_input(test_input, input_size))',
+          '        finish = time.time() - start',
+          '        time_array.append(finish)',
+          '    return time_array',
+          '    if time_array[1] > 30 * time_array[0]:',
+          '        return "not linear"',
+          '    return "linear"',
+          ''
+        ].join('\n');
+        console.log(test);
+        testCode += '\n' + [
+          VARNAME_TEST_RESULTS + '.append(',
+          '    run_performance_test(' + jsonVariableToPython(test.getInputDataAtom()) + '))'
+        ].join('\n');
+      });
 
-        var firstLine = 'class ' + WRAPPER_CLASS_NAME + '(object):';
-        var subsequentLines = codeLines.map(function(line) {
-          if (line.indexOf('def') === 0) {
-            var leftParenIndex = line.indexOf('(');
-            if (leftParenIndex === -1) {
-              throw Error('Incomplete line: missing "(" in def statement.');
-            }
-            return (
-              START_INDENT +
-              line.slice(0, leftParenIndex + 1) +
-              'self, ' +
-              line.slice(leftParenIndex + 1));
-          } else {
-            return START_INDENT + line;
-          }
-        });
+      return testCode;
+    };
 
-        var newCodeLines = [firstLine].concat(subsequentLines);
-        return newCodeLines.join('\n');
+    return {
+      preprocessCode: function(
+      code, mainFunctionName, correctnessTests, performanceTests) {
+        return (
+          wrapCodeIntoClass(code) + '\n' +
+          generateTestCode(
+              mainFunctionName, correctnessTests, performanceTests));
+      },
+      preprocessCorrectnessCode: function(code, mainFunctionName, correctnessTests) {
```

Comment:
Actually I'm not even sure why we need these extra functions at this level. Are they called? (I know you call them in the dispatcher but I don't think anything calls _those_ functions.)

You might want to expose them here with an underscore 'private' indicator for testing purposes (like I did for _wrapCodeIntoClass), but I don't think these should be part of the service's API.